drivers/apa102/apa102.c

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright (C) 2017 Freie Universität Berlin
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     drivers_apa102
+ * @{
+ *
+ * @file
+ * @brief       APA 102 RGB LED driver implementation
+ *
+ * @author      Hauke Petersen <hauke.petersen@fu-berlin.de>
+ *
+ * @}
+ */
+
+#include <string.h>
+
+#include "assert.h"
+#include "apa102.h"
+
+#define START           (0x00000000)
+#define END             (0xffffffff)
+
+#define HEAD            (0xe0000000)
+#define BRIGHT          (0x1f000000)
+#define BLUE            (0x00ff0000)
+#define GREEN           (0x0000ff00)
+#define RED             (0x000000ff)
+#define BRIGHT_SHIFT    (21U)
+#define BLUE_SHIFT      (16U)
+#define GREEN_SHIFT     (8U)
+
+
+
+static void shift(apa102_t *dev, uint32_t data)
+{
+    for (int i = 31; i >= 0; i--) {
+        gpio_write(dev->data_pin, ((data >> i) & 0x01));
+        gpio_set(dev->clk_pin);
+        gpio_clear(dev->clk_pin);
```

Comment:
nope, frquency is as fast as the functions are called (and as fast as the GPIOs are actually reacting). This could in theory up to 50MHz on some platforms, but this value is rather theoretical... On the samr21-xpro for example the actual clock rate is something around ~470KHz. Also the datasheet is quite blurry on the subject of maximum achievable clock rate, so I figured this solution should do...